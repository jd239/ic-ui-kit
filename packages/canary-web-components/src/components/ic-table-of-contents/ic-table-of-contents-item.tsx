import { FunctionalComponent, h } from "@stencil/core";

interface IcTableOfContentsItemElement {
  id: string;
  className: string;
  label: string;
  target: string;
  variant: string;
  selected: number;
  position: number;
  size?: string;
  truncate?: boolean;
  updateSelected: (position: number) => void;
  setItemLevelOffset: (variant: string) => string;
}

interface TOCItemProps {
  id: string;
  className: string;
  variant: string;
  label: string;
  target: string;
  selected: number;
  position: number;
  size?: string;
  truncate?: boolean;
  updateSelected: (position: number) => void;
}

const ItemHook = () => {
  // adds offset to the item based on its variant
  // e.g. h2 = 4 + (8*(2-2)) = 4px
  // h3 = 4 + (8*(3-2)) = 12px
  const setItemLevelOffset = (variant: string): string => {
    return `${4 + 8 * (parseInt(variant.split("")[1]) - 2)}px`;
  };

  return { setItemLevelOffset };
};

const ItemComponent: FunctionalComponent<IcTableOfContentsItemElement> = (
  itemProps
) => {
  const {
    id,
    className,
    label,
    target,
    selected,
    truncate,
    variant,
    position,
    size,
    updateSelected,
    setItemLevelOffset,
  } = itemProps;

  return (
    <a
      id={id}
      onClick={() => updateSelected(position)}
      class={`${selected && "table-of-contents-item-content-selected"}
        ${
          size === "large"
            ? `table-of-contents-item-content-large`
            : `table-of-contents-item-content-default`
        }
        ${className}
       `}
      href={`#${target}`}
    >
      <ic-typography
        id={`${id}-label`}
        bold={selected === position}
        class={`${truncate && "truncated-item-text"}  ${
          size && `table-of-contents-item-content-${size}`
        }
          table-of-contents-item-label
          `}
        style={{
          paddingLeft: setItemLevelOffset(variant),
        }}
      >
        {label}
      </ic-typography>
    </a>
  );
};

export const TableOfContentsItem = (props: TOCItemProps) => {
  const innerProps = ItemHook();
  return <ItemComponent {...props} {...innerProps} />;
};
