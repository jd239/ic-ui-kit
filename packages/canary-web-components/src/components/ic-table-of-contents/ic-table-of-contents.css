:host {
  display: block;
  border: none;
  max-height: 100vh;
}

.table-of-contents-section-container {
  display: flex;
}

.table-of-contents-content {
  top: 0;
  left: 0;
  padding: 0 var(--ic-space-md);
  font: var(--ic-color-text-primary);
  max-height: 100vh;
  z-index: 1;
}

:host .table-of-contents-position-top {
  min-width: 10rem;
  max-width: 50rem;
}

:host .table-of-contents-position-side {
  min-width: 8rem;
  max-width: 20rem;
  width: 20rem;
  position: sticky;
}

.table-of-contents-list {
  display: flex;
  flex-direction: row;
  max-width: inherit;
}

:host ol {
  list-style: none;
  margin-block-start: 0;
  padding-inline-start: 0;
}

.table-of-contents-list-items {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.table-of-contents-sidebar {
  background-color: var(--ic-table-of-contents-banner);
  height: 1.75rem;
  min-width: 0.25rem;
}

.table-of-contents-sidebar-monochrome {
  background-color: var(--ic-table-of-contents-banner-monochrome);
}

.table-of-contents-item {
  min-height: 1.5rem;
  line-height: 1.5rem;
}

.table-of-contents-item-content:hover {
  background-color: var(--ic-table-of-contents-hover);
}

.table-of-contents-item-monochrome-content:hover {
  background-color: var(--ic-table-of-contents-hover-monochrome);
}

.table-of-contents-item-content {
  border: none;
  max-width: inherit;
  display: flex;
  align-items: center;
  text-decoration: none;
  position: relative;
  color: var(--ic-table-of-contents-text);
  min-height: 1.5rem;

  --ic-typography-color: var(--ic-table-of-contents-text);
}

.table-of-contents-item-content-default {
  padding-top: var(--ic-space-xxxs);
  padding-bottom: var(--ic-space-xxxs);
}

.table-of-contents-item-content:focus {
  box-shadow: var(--ic-border-focus);
  border-radius: var(--ic-border-radius);
  transition: var(--ic-transition-duration-fast);
  outline: none;
  background-color: var(--ic-table-of-contents-hover);
}

.table-of-contents-item-content:active {
  box-shadow: var(--ic-border-focus);
  border-radius: var(--ic-border-radius);
  transition: var(--ic-transition-duration-fast);
  outline: none;
  background-color: var(--ic-table-of-contents-pressed);
}

.truncated-item-text {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

/* Large styles */

.table-of-contents-item-content-large {
  padding-top: var(--ic-space-xxs);
  padding-bottom: var(--ic-space-xxs);
}

.table-of-contents-sidebar.large {
  height: 2.5rem;
}

.table-of-contents-heading.large {
  padding-bottom: var(--ic-space-xs);
}

@media (prefers-reduced-motion: no-preference) {
  .table-of-contents-sidebar {
    transition: all 0.3s ease;
  }
}

@media (prefers-reduced-motion: reduce) {
  /* limit sidebar scroll animation */
  .table-of-contents-sidebar {
    transition: none;
  }
}

@media (forced-colors: active) {
  .table-of-contents-sidebar {
    background-color: Highlight;
  }
}
