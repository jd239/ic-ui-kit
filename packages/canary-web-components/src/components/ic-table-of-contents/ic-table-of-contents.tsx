import { Component, Element, h, Host, Prop, State, Watch } from "@stencil/core";
import { IcPositionTopOrRight, IcSizes, IcThemeMode } from "../../utils/types";
import { TableOfContentsItem } from "./ic-table-of-contents-item";
import { DEVICE_SIZES } from "../../utils/helpers";

const TOC_CONTENT_CLASS = "table-of-contents-content";
const TOC_LIST_ITEMS_CLASS = "table-of-contents-list-items";
const TOC_ITEM_CLASS = "table-of-contents-item-content";
const TOC_CONTAINER_CLASS = "table-of-contents-section-container";

interface RankedElement extends Element {
  rank: string;
  textLabel: string;
  treeChildren: RankedElement[];
  trueIndex: number;
}

@Component({
  tag: "ic-table-of-contents",
  styleUrl: "ic-table-of-contents.css",
  shadow: true,
})
export class TableOfContents {
  @Element() el: HTMLIcTableOfContentsElement;

  /**
   * Heading text of the table of contents.
   */
  @Prop() heading: string = "Contents";

  /**
   * Maximum heading level of child content that will generate an item in the table of contents. E.g. when set to '3', headings of <h4> and <h5> will not appear.
   */
  @Prop() maximumHeadingLevel: number = 3;

  /**
   * If `true`, the table of contents will display as black in the light theme, and white in dark theme.
   */
  @Prop() monochrome?: boolean = false;

  /**
   * Whether the table of contents will appear on the right of or above child content on large viewport widths.
   */
  @Prop({ mutable: true }) position: IcPositionTopOrRight = "right";

  /**
   * Vertical offset to page scroll (in percent) when table of contents item is selected, to account for other page components like headers.
   */
  @Prop() setScrollHeight: number = 50;

  /**
   * The size of the table of contents items.
   */
  @Prop() size: IcSizes = "medium";

  /**
   * Sets the theme color to the dark or light theme color. "inherit" will set the color based on the system settings or ic-theme component.
   */
  @Prop() theme?: IcThemeMode = "inherit";

  /**
   * If `true`, the table of contents item label will be truncated.
   */
  @Prop() truncate: boolean = true;

  @State() focusedElement: number = 0;

  // heading items
  private referenceElements: HTMLElement[] = [];

  // heights of heading items
  private currentTargetHeights: number[] = [];

  private resizeObserver: ResizeObserver;
  // header height if present
  private headerHeight: number = 0;
  // holds initial position to allow for responsive changes
  private initialPosition: IcPositionTopOrRight = this.position;

  // suppress scroll listener to allow for click selection without scroll listener firing
  @State() suppressListener: boolean = false;

  @Watch("suppressListener")
  suppressListenerHandler(): void {
    window.scrollTo(0, window.scrollY - this.headerHeight - 50);
  }

  /**
   * Generate a TOC Item.
   * These items are wrapped in empty tooltips that will be populated by addTooltip()
   * @param tocHeading heading element to refer item information from
   * @param children
   * @returns TocItem ready to be rendered, with nested children if applicable
   */
  private createTocItem = (
    tocHeading: RankedElement,
    children?: HTMLElement[]
  ): HTMLElement => {
    const tocId = `toc_${tocHeading.id}`;

    const tocItem = (
      <TableOfContentsItem
        id={tocId}
        className={TOC_ITEM_CLASS}
        variant={tocHeading.tagName.toLowerCase()}
        label={tocHeading.textLabel}
        target={tocHeading.id}
        truncate={this.truncate}
        selected={this.focusedElement}
        updateSelected={this.updateOnSelection}
        position={tocHeading.trueIndex}
        size={this.size}
      />
    );

    const wrappedToc = this.truncate ? (
      <ic-tooltip target={tocId} label={tocHeading.innerHTML}>
        {tocItem}
      </ic-tooltip>
    ) : (
      tocItem
    );
    return (
      <li
        class={`table-of-contents-item  ${
          this.monochrome && "table-of-contents-item-monochrome"
        }`}
      >
        {wrappedToc}
        {children ? <ol>{children}</ol> : null}
      </li>
    );
  };

  /**
   * Render TocItems and recursively render children items if they exist
   * @param listItem
   * @returns HTML Element array to be rendered
   */
  private renderTocTree = (listItem: RankedElement): HTMLElement => {
    let results;
    if (listItem.treeChildren.length > 0) {
      const children = listItem.treeChildren.map((i) => this.renderTocTree(i));
      results = this.createTocItem(listItem, children);
    } else {
      results = this.createTocItem(listItem);
    }
    return results;
  };

  /**
   * Scrape for heading elements on the page content within TOC, down to the maximumHeadingLevel.
   * From this, generate an array of Elements with lower rank headings inserted as treeChildren on top level headings.
   * @returns RankedElement[] array of HTML elements with additional props for rank, trueIndex and treeChildren array
   */
  private generateHeadingTree(): RankedElement[] {
    const resultArray: RankedElement[] = [];

    const headingDepth =
      this.maximumHeadingLevel < 5
        ? ["h2", "h3", "h4", "h5"]
            .slice(
              0,
              (this.maximumHeadingLevel < 2 ? 2 : this.maximumHeadingLevel) - 1
            )
            .join(",")
        : "h2, h3, h4, h5";
    const headingList = this.el.querySelectorAll(headingDepth);

    if (headingList) {
      const headingArray = Array.from(headingList) as RankedElement[];
      headingArray.forEach((item, index) => {
        this.referenceElements.push(item as unknown as HTMLElement);
        item.trueIndex = index;
        item.rank = item.tagName.toLowerCase().slice(1, 2);
        item.textLabel = item.textContent
          ? item.textContent.trimStart().split("\n")[0]
          : `${item.rank}-${index}`;
        item.treeChildren = [];
      });

      let currentRank = headingArray[0]?.rank || 2;

      headingArray.reverse().forEach((h, i) => {
        if (!h.id) {
          h.setAttribute("id", h.textLabel.replace(/\s+/g, "-").toLowerCase());
        }
        if (h.rank > currentRank && h.trueIndex > 0) {
          if (headingArray[i] && headingArray[i].rank < h.rank) {
            headingArray[i].treeChildren.unshift(h);
          } else {
            resultArray.push(h);
          }
        } else {
          currentRank = h.rank;
          resultArray.push(h);
        }
      });
    }
    return resultArray.reverse();
  }

  //container orientation logic
  private containerPosition(): void {
    if (this.position) {
      const container = this.el.shadowRoot?.querySelector(
        `.${TOC_CONTAINER_CLASS}`
      ) as HTMLIcSectionContainerElement;
      switch (this.position) {
        case "right":
          container.style.flexDirection = "row-reverse";
          break;
        case "top":
          container.style.flexDirection = "column";
          break;
      }
    }
  }

  /**
   * Move the sidebar to the provided position on the page.
   * Also, resize the sidebar to stretch the height of the selected heading.
   * @param newPosition
   */
  private sidebarTranslation(newPosition: number): void {
    const sidebar = this.el.shadowRoot?.querySelector(
      ".table-of-contents-sidebar"
    ) as HTMLIcSectionContainerElement;

    const itemHeights: number[] = [];
    let previousItemsHeight = 0;
    this.el.shadowRoot
      ?.querySelectorAll(`.${TOC_ITEM_CLASS}`)
      ?.forEach((item: any) => {
        if (item !== null) itemHeights.push(item.offsetHeight);
      });
    itemHeights
      .slice(0, newPosition)
      .forEach((i) => (previousItemsHeight += i));

    sidebar.style.transform = `translateY(${previousItemsHeight}px)`;
    sidebar.style.height = `${itemHeights[newPosition]}px`;
  }

  // list item clicked logic
  private updateOnSelection = (position: number) => {
    this.suppressListener = true;
    this.focusedElement = position;
    this.sidebarTranslation(this.focusedElement);
  };

  // find element heights to be used to determine focused item from scroll position
  private findElementsHeights() {
    const contentBounds: number[] = [];
    this.referenceElements.forEach((element) => {
      contentBounds.push(element.getBoundingClientRect().top);
    });
    this.currentTargetHeights = contentBounds;
  }

  //resize window logic
  private handleResizeBounds() {
    if (window.innerWidth < DEVICE_SIZES.XL && this.initialPosition !== "top") {
      this.position = "top";
      this.containerPosition();
    } else if (
      window.innerWidth > DEVICE_SIZES.XL &&
      this.position === "top" &&
      this.initialPosition === "right"
    ) {
      this.position = "right";
      this.containerPosition();
    }
    this.findElementsHeights();
    this.truncateText();
  }

  // scroll logic
  private updateSelectedItemFromScroll() {
    const targetHeight =
      window.scrollY + window.innerHeight * (this.setScrollHeight / 100);
    if (window.scrollY === 0) {
      this.focusedElement = 0;
      this.sidebarTranslation(this.focusedElement);
    } else if (targetHeight < this.currentTargetHeights[this.focusedElement]) {
      const previousItems = this.currentTargetHeights.slice(
        0,
        this.focusedElement
      );
      const closestItem = previousItems.reduce((newValue, curr, index) => {
        if (
          Math.abs(curr - targetHeight) <
          Math.abs(previousItems[newValue] - targetHeight)
        )
          return index;
        return newValue;
      }, 0);
      this.focusedElement = closestItem;
      this.sidebarTranslation(this.focusedElement);
    } else if (
      targetHeight > this.currentTargetHeights[this.focusedElement] &&
      this.focusedElement !== this.currentTargetHeights.length
    ) {
      const relevantItems = this.currentTargetHeights.slice(
        this.focusedElement
      );

      const closestItem = relevantItems.reduce((newValue, curr, index) => {
        if (
          Math.abs(curr - targetHeight) <
            Math.abs(relevantItems[newValue] - targetHeight) &&
          curr < targetHeight
        )
          return index;
        return newValue;
      }, 0);
      if (closestItem + this.focusedElement !== this.focusedElement) {
        this.focusedElement += closestItem;
        this.sidebarTranslation(this.focusedElement);
      }
    }
  }

  // find header element height
  private findHeaderElement(): number {
    const element = document.querySelector("ic-page-header");
    if (element) {
      return element.clientHeight;
    }
    const header = document.querySelector("header");
    if (header) {
      return header.clientHeight;
    }
    return 0;
  }

  // add margin to the list based off the header element height
  private addListMargin(): void {
    const content = this.el.shadowRoot?.querySelector(
      `.${TOC_CONTENT_CLASS}`
    ) as HTMLDivElement;
    content?.style.setProperty("top", `${this.headerHeight}px`);
  }

  // add tooltip to the item if it is too wide
  private addTooltip(item: any, widthBreakpoint: number) {
    const itemLabel = item.querySelector(".table-of-contents-item-label");
    if (item.parentNode?.tagName == "IC-TOOLTIP") {
      const tooltip = item.parentNode;
      if (itemLabel?.offsetWidth < itemLabel?.scrollWidth) {
        const tooltipPosition =
          window.innerWidth < widthBreakpoint
            ? "bottom"
            : this.position === "right"
            ? "left"
            : "right";
        //Modify the ic-tooltip wrapped around the list item
        tooltip.setAttribute("disable-click", false);
        tooltip.setAttribute("disable-hover", false);
        tooltip.setAttribute("placement", tooltipPosition);
        tooltip.setAttribute("target", item.id);
        tooltip.setAttribute("style", "overflow:hidden");
      } else {
        //disable the tooltip if the item does not need to truncate
        tooltip.setAttribute("disable-click", true);
        tooltip.setAttribute("disable-hover", true);
      }
    }
  }

  // truncate text if it goes out of bounds of the container
  private truncateText = () => {
    const links = this.el.shadowRoot?.querySelectorAll(`.${TOC_ITEM_CLASS}`);
    const contentList = this.el.shadowRoot?.querySelector(
      `.${TOC_LIST_ITEMS_CLASS}`
    );
    if (contentList)
      links?.forEach((item) => {
        this.addTooltip(item, DEVICE_SIZES.S);
      });
  };

  // initialise resize observer
  componentWillLoad(): void {
    this.resizeObserver = new ResizeObserver(() => this.handleResizeBounds());
    this.resizeObserver.observe(this.el);
  }

  // initialise scroll listener
  componentDidRender(): void {
    document.addEventListener("scroll", () => {
      if (!this.suppressListener) this.updateSelectedItemFromScroll();
      else this.suppressListener = false;
    });
    this.containerPosition();
    this.findElementsHeights();
    this.sidebarTranslation(this.focusedElement);
  }

  // after component has loaded modify the list based of position, size and surrounding elements
  async componentDidLoad(): Promise<void> {
    this.focusedElement = 0;
    this.headerHeight = this.findHeaderElement();
    this.addListMargin();
    this.truncateText();
    this.findElementsHeights();
    this.sidebarTranslation(this.focusedElement);
  }

  // Component is a wrapper for page contents, and will build list from heading items
  render() {
    const { heading, monochrome, position, size, theme } = this;
    return (
      <Host class={{ [`ic-theme-${theme}`]: theme !== "inherit" }}>
        <ic-section-container class={TOC_CONTAINER_CLASS} aligned="full-width">
          <div
            role="navigation"
            class={`${TOC_CONTENT_CLASS} ${
              position === "top" && "table-of-contents-position-top"
            }
            ${position === "right" && "table-of-contents-position-side"}`}
          >
            <ic-typography
              class={`table-of-contents-heading ${size}`}
              variant="subtitle-large"
            >
              {heading}
            </ic-typography>
            <div class="table-of-contents-list">
              <div
                class={`table-of-contents-sidebar ${
                  monochrome && `table-of-contents-sidebar-monochrome`
                }
                ${size}
                `}
              />
              <ol class={TOC_LIST_ITEMS_CLASS}>
                {this.generateHeadingTree().map((item) =>
                  this.renderTocTree(item)
                )}
              </ol>
            </div>
          </div>
          <slot></slot>
        </ic-section-container>
      </Host>
    );
  }
}
