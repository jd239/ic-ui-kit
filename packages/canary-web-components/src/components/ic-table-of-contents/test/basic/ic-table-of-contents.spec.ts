import { newSpecPage, SpecPage } from "@stencil/core/testing";
import { TableOfContents } from "../../ic-table-of-contents";

describe("ic-table-of-contents", () => {
  it("should render default", async () => {
    const page = await newSpecPage({
      components: [TableOfContents],
      html: `<ic-table-of-contents>
      <div><h1>Heading</h1></div>
      </ic-table-of-contents>
        `,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render default top", async () => {
    const page = await newSpecPage({
      components: [TableOfContents],
      html: `<ic-table-of-contents position="top">
      <div><h1>Heading</h1></div>
      </ic-table-of-contents>
        `,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render default with multiple headings", async () => {
    const page = await newSpecPage({
      components: [TableOfContents],
      html: `<ic-table-of-contents>
      <div>
      <h1>Heading</h1>
        <h2>Subheading</h2>
      </div>
      </ic-table-of-contents>
        `,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render default with custom title", async () => {
    const page = await newSpecPage({
      components: [TableOfContents],
      html: `<ic-table-of-contents heading="Custom Title">
      <div>
      <h1>Heading</h1>
        <h2>Subheading</h2>
      </div>
      </ic-table-of-contents>
        `,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render maximum heading depth", async () => {
    const page = await newSpecPage({
      components: [TableOfContents],
      html: `<ic-table-of-contents heading="Custom Title">
      <div>
      <h1>Heading</h1>
        <h2>Subheading</h2>
        <h3>Sub-subheading</h3>
        <h4>Sub-sub-subheading</h4>
        <h5>Sub-sub-sub-subheading</h5>
      </div>
      </ic-table-of-contents>
        `,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with correct targets", async () => {
    await newSpecPage({
      components: [TableOfContents],
      html: `<ic-table-of-contents>
      <div>
        <h2 id="subheading1">Subheading 1</h2>
        <h3 id="subsubheading1">Sub-subheading 1</h3>
      </div>
      </ic-table-of-contents>
        `,
    });

    const toc = document.querySelector(
      "ic-table-of-contents"
    ) as HTMLIcTableOfContentsElement;
    expect(toc).toBeTruthy();

    const tocItems = toc.shadowRoot?.querySelectorAll(
      ".table-of-contents-item"
    ) as NodeListOf<Element>;
    expect(tocItems.length).toBe(2);
    const subheading1 = tocItems[0].querySelector("a") as HTMLAnchorElement;
    expect(subheading1.getAttribute("href")).toBe("#subheading1");
    window.scrollTo = jest.fn();
    subheading1.click();
    expect(window.scrollTo).toHaveBeenCalledTimes(1);
  });

  it("should respond to scroll events", async () => {
    const page = await newSpecPage({
      components: [TableOfContents],
      html: `<ic-table-of-contents>
      <div>
      <h2>Heading</h2>
      <h2>Heading 2</h2>
        <h3>Subheading</h3>
      </div>
      </ic-table-of-contents>
        `,
    });

    const toc = document.querySelector(
      "ic-table-of-contents"
    ) as HTMLIcTableOfContentsElement;
    expect(toc).toBeTruthy();

    const tocItems = toc.shadowRoot?.querySelectorAll(
      ".table-of-contents-item"
    ) as NodeListOf<Element>;
    expect(tocItems.length).toBe(3);

    page.rootInstance.currentTargetHeights = [0, 100, 600];
    page.rootInstance.sidebarTranslation = jest.fn();

    expect(window.scrollY).toBe(0);
    document.dispatchEvent(new Event("scroll"));
    expect(page.rootInstance.sidebarTranslation).toHaveBeenCalledWith(0);

    window.scrollY = 500;
    document.dispatchEvent(new Event("scroll"));
    expect(page.rootInstance.sidebarTranslation).toHaveBeenCalledWith(2);
    expect(page.rootInstance.focusedElement).toBe(2);

    window.scrollY = 100;
    document.dispatchEvent(new Event("scroll"));
    expect(page.rootInstance.sidebarTranslation).toHaveBeenCalledWith(1);
    expect(page.rootInstance.focusedElement).toBe(1);
  });

  it("should render correctly with header present", async () => {
    await newSpecPage({
      components: [TableOfContents],
      html: `<ic-page-header
       heading="Latte recipe"
            subheading="A Latte is a popular Italian coffee, made with espresso, steamed milk and a thin layer of foam."
            reverse-order="true"
            sticky="true"></ic-page-header>
      <ic-table-of-contents heading="Custom Title">
      <div>
      <h1>Heading</h1>
        <h2>Subheading</h2>
        <h3>Sub-subheading</h3>
        <h4>Sub-sub-subheading</h4>
        <h5>Sub-sub-sub-subheading</h5>
      </div>
      </ic-table-of-contents>
        `,
    });

    const pageHeader = document.querySelector(
      "ic-page-header"
    ) as HTMLIcPageHeaderElement;

    expect(pageHeader).toBeTruthy();
  });

  it("should render correctly with ic-header present", async () => {
    await newSpecPage({
      components: [TableOfContents],
      html: `<header>Header</header>
      <ic-table-of-contents heading="Custom Title">
      <div>
      <h1>Heading</h1>
        <h2>Subheading</h2>
        <h3>Sub-subheading</h3>
        <h4>Sub-sub-subheading</h4>
        <h5>Sub-sub-sub-subheading</h5>
      </div>
      </ic-table-of-contents>
        `,
    });

    const pageHeader = document.querySelector("header") as HTMLElement;

    expect(pageHeader).toBeTruthy();
  });

  it("should truncate text correctly", async () => {
    const page: SpecPage = await newSpecPage({
      components: [TableOfContents],
      html: `
      <ic-table-of-contents heading="Custom Title" truncate="true">
      <div>
        <h2 id="test">Heading with long message that should be truncated</h2>
      </div> 
      </ic-table-of-contents>
        `,
    });
    const toc = document.querySelector(
      "ic-table-of-contents"
    ) as HTMLIcTableOfContentsElement;
    expect(toc).toBeTruthy();
    const tocItems = toc.shadowRoot?.querySelectorAll(
      ".table-of-contents-item"
    ) as NodeListOf<Element>;
    expect(tocItems.length).toBe(1);

    page.rootInstance.addTooltip(tocItems[0], 400, 320);
    const tooltip = toc.shadowRoot?.querySelector("ic-tooltip");
    expect(tooltip).not.toBeNull();
  });

  it("should change position based on page width", async () => {
    const page: SpecPage = await newSpecPage({
      components: [TableOfContents],
      html: `
      <ic-table-of-contents heading="Custom Title" truncate="true">
      <div>
        <h2 id="test">Heading with long message that should be truncated</h2>
      </div> 
      </ic-table-of-contents>
        `,
    });
    expect(page.rootInstance.position).toBe("right");
    window.innerWidth = 900;
    page.rootInstance.handleResizeBounds();
    expect(page.rootInstance.position).toBe("top");
    window.innerWidth = 1300;
    page.rootInstance.handleResizeBounds();
    expect(page.rootInstance.position).toBe("right");
  });

  it("should suppress scroll listener", async () => {
    const page: SpecPage = await newSpecPage({
      components: [TableOfContents],
      html: `
      <ic-table-of-contents heading="Custom Title" truncate="true">
      <div>
        <h2 id="test">Heading with long message that should be truncated</h2>
      </div> 
      </ic-table-of-contents>
        `,
    });
    page.rootInstance.updateSelectedItemFromScroll = jest.fn();
    expect(page.rootInstance.updateSelectedItemFromScroll).toBeCalledTimes(0);
    page.rootInstance.suppressListener = true;
    document.dispatchEvent(new Event("scroll"));
    expect(page.rootInstance.updateSelectedItemFromScroll).toBeCalledTimes(0);
  });

  it("should apply correct styling for monochrome", async () => {
    await newSpecPage({
      components: [TableOfContents],
      html: `
      <ic-table-of-contents heading="Custom Title" monochrome="true">
      <div>
        <h2 id="test">Heading with long message that should be truncated</h2>
      </div> 
      </ic-table-of-contents>
        `,
    });
    const toc = document.querySelector(
      "ic-table-of-contents"
    ) as HTMLIcTableOfContentsElement;
    expect(toc).toBeTruthy();
    const tocItems = toc.shadowRoot?.querySelectorAll(
      ".table-of-contents-item"
    ) as NodeListOf<Element>;
    expect(tocItems.length).toBe(1);

    expect(
      tocItems[0].classList.contains("table-of-contents-item-monochrome")
    ).toBe(true);
  });
});
