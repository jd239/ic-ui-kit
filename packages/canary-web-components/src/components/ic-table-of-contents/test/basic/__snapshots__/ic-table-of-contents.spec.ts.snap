// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-table-of-contents should render default 1`] = `
<ic-table-of-contents>
  <template shadowrootmode="open">
    <ic-section-container aligned="full-width" class="table-of-contents-section-container" style="flex-direction: row-reverse;">
      <div class="false table-of-contents-content table-of-contents-position-side" role="navigation" style="top: 0px;">
        <ic-typography class="medium table-of-contents-heading" variant="subtitle-large">
          Contents
        </ic-typography>
        <div class="table-of-contents-list">
          <div class="false medium table-of-contents-sidebar"></div>
          <ol class="table-of-contents-list-items"></ol>
        </div>
      </div>
      <slot></slot>
    </ic-section-container>
  </template>
  <div>
    <h1>
      Heading
    </h1>
  </div>
</ic-table-of-contents>
`;

exports[`ic-table-of-contents should render default top 1`] = `
<ic-table-of-contents position="top">
  <template shadowrootmode="open">
    <ic-section-container aligned="full-width" class="table-of-contents-section-container" style="flex-direction: column;">
      <div class="false table-of-contents-content table-of-contents-position-top" role="navigation" style="top: 0px;">
        <ic-typography class="medium table-of-contents-heading" variant="subtitle-large">
          Contents
        </ic-typography>
        <div class="table-of-contents-list">
          <div class="false medium table-of-contents-sidebar"></div>
          <ol class="table-of-contents-list-items"></ol>
        </div>
      </div>
      <slot></slot>
    </ic-section-container>
  </template>
  <div>
    <h1>
      Heading
    </h1>
  </div>
</ic-table-of-contents>
`;

exports[`ic-table-of-contents should render default with custom title 1`] = `
<ic-table-of-contents heading="Custom Title">
  <template shadowrootmode="open">
    <ic-section-container aligned="full-width" class="table-of-contents-section-container" style="flex-direction: row-reverse;">
      <div class="false table-of-contents-content table-of-contents-position-side" role="navigation" style="top: 0px;">
        <ic-typography class="medium table-of-contents-heading" variant="subtitle-large">
          Custom Title
        </ic-typography>
        <div class="table-of-contents-list">
          <div class="false medium table-of-contents-sidebar"></div>
          <ol class="table-of-contents-list-items">
            <li class="false table-of-contents-item">
              <ic-tooltip disable-click="true" disable-hover="true" label="Subheading" target="toc_subheading">
                <a class="0 table-of-contents-item-content table-of-contents-item-content-default" href="#subheading" id="toc_subheading">
                  <ic-typography bold="" class="table-of-contents-item-content-medium table-of-contents-item-label truncated-item-text" id="toc_subheading-label" style="padding-left: 4px;">
                    Subheading
                  </ic-typography>
                </a>
              </ic-tooltip>
            </li>
          </ol>
        </div>
      </div>
      <slot></slot>
    </ic-section-container>
  </template>
  <div>
    <h1>
      Heading
    </h1>
    <h2 id="subheading">
      Subheading
    </h2>
  </div>
</ic-table-of-contents>
`;

exports[`ic-table-of-contents should render default with multiple headings 1`] = `
<ic-table-of-contents>
  <template shadowrootmode="open">
    <ic-section-container aligned="full-width" class="table-of-contents-section-container" style="flex-direction: row-reverse;">
      <div class="false table-of-contents-content table-of-contents-position-side" role="navigation" style="top: 0px;">
        <ic-typography class="medium table-of-contents-heading" variant="subtitle-large">
          Contents
        </ic-typography>
        <div class="table-of-contents-list">
          <div class="false medium table-of-contents-sidebar"></div>
          <ol class="table-of-contents-list-items">
            <li class="false table-of-contents-item">
              <ic-tooltip disable-click="true" disable-hover="true" label="Subheading" target="toc_subheading">
                <a class="0 table-of-contents-item-content table-of-contents-item-content-default" href="#subheading" id="toc_subheading">
                  <ic-typography bold="" class="table-of-contents-item-content-medium table-of-contents-item-label truncated-item-text" id="toc_subheading-label" style="padding-left: 4px;">
                    Subheading
                  </ic-typography>
                </a>
              </ic-tooltip>
            </li>
          </ol>
        </div>
      </div>
      <slot></slot>
    </ic-section-container>
  </template>
  <div>
    <h1>
      Heading
    </h1>
    <h2 id="subheading">
      Subheading
    </h2>
  </div>
</ic-table-of-contents>
`;

exports[`ic-table-of-contents should render maximum heading depth 1`] = `
<ic-table-of-contents heading="Custom Title">
  <template shadowrootmode="open">
    <ic-section-container aligned="full-width" class="table-of-contents-section-container" style="flex-direction: row-reverse;">
      <div class="false table-of-contents-content table-of-contents-position-side" role="navigation" style="top: 0px;">
        <ic-typography class="medium table-of-contents-heading" variant="subtitle-large">
          Custom Title
        </ic-typography>
        <div class="table-of-contents-list">
          <div class="false medium table-of-contents-sidebar"></div>
          <ol class="table-of-contents-list-items">
            <li class="false table-of-contents-item">
              <ic-tooltip disable-click="true" disable-hover="true" label="Subheading" target="toc_subheading">
                <a class="0 table-of-contents-item-content table-of-contents-item-content-default" href="#subheading" id="toc_subheading">
                  <ic-typography bold="" class="table-of-contents-item-content-medium table-of-contents-item-label truncated-item-text" id="toc_subheading-label" style="padding-left: 4px;">
                    Subheading
                  </ic-typography>
                </a>
              </ic-tooltip>
            </li>
            <li class="false table-of-contents-item">
              <ic-tooltip disable-click="true" disable-hover="true" label="Sub-subheading" target="toc_sub-subheading">
                <a class="0 table-of-contents-item-content table-of-contents-item-content-default" href="#sub-subheading" id="toc_sub-subheading">
                  <ic-typography class="table-of-contents-item-content-medium table-of-contents-item-label truncated-item-text" id="toc_sub-subheading-label" style="padding-left: 12px;">
                    Sub-subheading
                  </ic-typography>
                </a>
              </ic-tooltip>
            </li>
          </ol>
        </div>
      </div>
      <slot></slot>
    </ic-section-container>
  </template>
  <div>
    <h1>
      Heading
    </h1>
    <h2 id="subheading">
      Subheading
    </h2>
    <h3 id="sub-subheading">
      Sub-subheading
    </h3>
    <h4>
      Sub-sub-subheading
    </h4>
    <h5>
      Sub-sub-sub-subheading
    </h5>
  </div>
</ic-table-of-contents>
`;
